File: frontend/src/app/graphs/FeatureImportance.js
import { Bar } from "react-chartjs-2";
import {
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    BarElement,
    Title,
    Tooltip,
    Legend,
} from "chart.js";

ChartJS.register(CategoryScale, LinearScale, BarElement, Title, Tooltip, Legend);
export default function FeatureImportance() {

    const features = {
        "Orbital Period (days)": 27884.833504080772,
        "Transit Duration (hours)": 11461.031258970499,
        "Transit Depth (ppm)": 20514.496178925037,
        "Impact Parameter": 9013.564484089613,
        "Eccentricity": 872.1251285076141,
        "Inclination (degrees)": 8752.685040086508,
        "Planet Radius (Earth radii)": 50285.66988945007,
        "Planet Equilibrium Temperature (K)": 25769.203072845936,
        "Insolation Flux (Earth flux)": 6548.60926130414,
        "Stellar Effective Temperature (K)": 8623.679917544127,
        "Stellar Surface Gravity (log10(cm/s^2))": 7475.145740032196,
        "Stellar Radius (solar radii)": 6567.19534599781,
        "Stellar Mass (solar masses)": 4982.842602550983,
        "Stellar Metallicity ([Fe/H])": 13078.731902450323,
        "Dataset Type Indicator": 3818.9925240278244
    }
    // Sort features by importance (descending)
    const sorted = Object.entries(features).sort((a, b) => b[1] - a[1]);
    const labels = sorted.map(([name]) => name);
    const values = sorted.map(([_, importance]) => importance);

    const data = {
        labels,
        datasets: [
            {
                label: "Importance",
                data: values,
                backgroundColor: "rgba(168, 85, 247, 0.8)", // purple
                borderColor: "rgba(168, 85, 247, 1)",
                borderWidth: 1,
            },
        ],
    };

    const options = {
        indexAxis: "y", // horizontal bar chart
        responsive: true,
        plugins: {
            legend: { display: false },
            title: {
                display: true,
                text: "Feature importance",
                color: "#fff",
                font: { size: 18 },
            },
        },
        scales: {
            x: {
                ticks: { color: "#fff" },
                grid: { color: "rgba(255,255,255,0.1)" },
            },
            y: {
                ticks: { color: "#fff" },
                grid: { color: "rgba(255,255,255,0.1)" },
            },
        },
    };

    return (
        <div className="relative w-full max-w-2xl p-6 rounded-xl bg-gradient-to-br from-[#1a1a2e] via-[#13111c] to-[#0d0b13] shadow-xl border border-white/10">
            <Bar data={data} options={options} />
        </div>
    );
}

File: astrum_ai/feature_importance.py
from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Dict, List

from joblib import load

from .common import FEATURE_COLUMNS, MODEL_FILENAME, PREPROCESSOR_FILENAME

_DEFAULT_OUTPUT_NAME = "feature_importances.json"


class ArtifactNotFoundError(FileNotFoundError):
    """Raised when a required model artifact is missing."""


# Human readable labels derived from the training feature definitions.
_SEMANTIC_NAME_MAP = {
    "orbital_period": "Orbital Period (days)",
    "transit_duration": "Transit Duration (hours)",
    "transit_depth": "Transit Depth (ppm)",
    "impact_parameter": "Impact Parameter",
    "eccentricity": "Eccentricity",
    "inclination": "Inclination (degrees)",
    "planet_radius": "Planet Radius (Earth radii)",
    "planet_equilibrium_temp": "Planet Equilibrium Temperature (K)",
    "insolation_flux": "Insolation Flux (Earth flux)",
    "stellar_temp": "Stellar Effective Temperature (K)",
    "stellar_logg": "Stellar Surface Gravity (log10(cm/s^2))",
    "stellar_radius": "Stellar Radius (solar radii)",
    "stellar_mass": "Stellar Mass (solar masses)",
    "stellar_metallicity": "Stellar Metallicity ([Fe/H])",
    "dataset_type_id": "Dataset Type Indicator",
}


def _load_shared_model(model_dir: Path):
    model_path = model_dir / MODEL_FILENAME
    if not model_path.exists():
        raise ArtifactNotFoundError(
            f"Shared model artifact not found at '{model_path}'. Train the model before computing importances."
        )
    return load(model_path)


def _load_feature_metadata(model_dir: Path) -> List[str]:
    bundle_path = model_dir / PREPROCESSOR_FILENAME
    if not bundle_path.exists():
        raise ArtifactNotFoundError(
            f"Preprocessing bundle not found at '{bundle_path}'. Train the model before computing importances."
        )
    bundle = load(bundle_path)
    feature_columns = bundle.get("feature_columns") or FEATURE_COLUMNS
    return [*feature_columns, "dataset_type_id"]


def _compute_importances(model, feature_names: List[str], importance_type: str) -> Dict[str, float]:
    booster = getattr(model, "booster_", None)
    if booster is None:
        raise RuntimeError("Loaded model does not expose a LightGBM booster; cannot compute feature importances.")

    importances = booster.feature_importance(importance_type=importance_type)
    booster_names = booster.feature_name() or []

    if len(importances) == len(feature_names):
        canonical_names = list(feature_names)
    elif booster_names and len(booster_names) == len(importances):
        canonical_names = list(booster_names)
    else:
        raise RuntimeError(
            "Mismatch between feature importances and expected feature names; "
            "retrain the model to refresh metadata."
        )

    semantic_names = [_SEMANTIC_NAME_MAP.get(name, name) for name in canonical_names]
    return {name: float(importance) for name, importance in zip(semantic_names, importances)}


def compute_feature_importances(
    model_dir: Path,
    output_path: Path | None = None,
    importance_type: str = "gain",
) -> Path:
    """Compute and persist LightGBM feature importances as JSON."""
    model_dir = model_dir.expanduser().resolve()
    model = _load_shared_model(model_dir)
    feature_names = _load_feature_metadata(model_dir)

    importance_type = importance_type.lower()
    valid_types = {"gain", "split"}
    if importance_type not in valid_types:
        raise ValueError(
            f"importance_type must be one of {sorted(valid_types)}, got '{importance_type}'."
        )

    importances = _compute_importances(model, feature_names, importance_type)

    output_path = (output_path or model_dir / _DEFAULT_OUTPUT_NAME).resolve()
    output_path.write_text(json.dumps(importances, indent=2, sort_keys=True))
    return output_path


def feature_importance_cli() -> None:
    parser = argparse.ArgumentParser(
        description="Compute LightGBM feature importances for the shared model and save them as JSON."
    )
    parser.add_argument(
        "--model-dir",
        default="assets/models",
        help="Directory containing the trained shared model artifacts (default: assets/models).",
    )
    parser.add_argument(
        "--output",
        help="Optional output path for the JSON file (default: <model-dir>/feature_importances.json).",
    )
    parser.add_argument(
        "--importance-type",
        default="gain",
        choices=["gain", "split"],
        help="Importance type to request from LightGBM (default: gain).",
    )

    args = parser.parse_args()

    model_dir = Path(args.model_dir)
    output_path = Path(args.output) if args.output else None

    result_path = compute_feature_importances(model_dir, output_path, args.importance_type)
    print(f"Feature importances saved to '{result_path}'.")


if __name__ == "__main__":
    feature_importance_cli()
